graph/network/diffusion_test.go

```diff
@@ -0,0 +1,495 @@
+				K: 0.0006161414, L: 0.0000148186, M: 0.0000001852, N: 0.0000000016, O: 0.0000000000,
+				P: 0.0000068453, Q: 0.0000001420, R: 0.0000000016, S: 0.0000000000, T: 0.0000000000,
+				U: 0.0000000699, V: 0.0000000014, W: 0.0000000000, X: 0.0000000000, Y: 0.0000000000,
+			},
+		},
+	},
+	{
+		g: grid(5),
+		h: map[int64]float64{0: 1},
+		t: 1,
+
+		wantTol: 1e-9,
+		want: map[bool]map[int64]float64{
+			false: {
+				A: 0.2743435076, B: 0.1615920872, C: 0.0639346641, D: 0.0188054933, E: 0.0051023569,
+				F: 0.1615920872, G: 0.0951799548, H: 0.0376583937, I: 0.0110766934, J: 0.0030053582,
+				K: 0.0639346641, L: 0.0376583937, M: 0.0148997194, N: 0.0043825455, O: 0.0011890840,
+				P: 0.0188054933, Q: 0.0110766934, R: 0.0043825455, S: 0.0012890649, T: 0.0003497525,
+				U: 0.0051023569, V: 0.0030053582, W: 0.0011890840, X: 0.0003497525, Y: 0.0000948958,
+			},
+			true: {
+				A: 0.4323917545, B: 0.1660487336, C: 0.0270298904, D: 0.0029720194, E: 0.0003007247,
+				F: 0.1660487336, G: 0.0463974679, H: 0.0063556078, I: 0.0006056850, J: 0.0000589574,
+				K: 0.0270298904, L: 0.0063556078, M: 0.0007860810, N: 0.0000691647, O: 0.0000065586,
+				P: 0.0029720194, Q: 0.0006056850, R: 0.0000691647, S: 0.0000057466, T: 0.0000005475,
+				U: 0.0003007247, V: 0.0000589574, W: 0.0000065586, X: 0.0000005475, Y: 0.0000000555,
+			},
+		},
+	},
+	{
+		g: grid(5),
+		h: map[int64]float64{0: 1},
+		t: 10,
+
+		wantTol: 1e-9,
+		want: map[bool]map[int64]float64{
+			false: {
+				A: 0.0432408511, B: 0.0425986522, C: 0.0415977802, D: 0.0405588482, E: 0.0399403788,
+				F: 0.0425986522, G: 0.0420083007, H: 0.0409532810, I: 0.0399982373, J: 0.0393463013,
+				K: 0.0415977802, L: 0.0409532810, M: 0.0400339958, N: 0.0389913353, O: 0.0384232854,
+				P: 0.0405588482, Q: 0.0399982373, R: 0.0389913353, S: 0.0380844049, T: 0.0374622025,
+				U: 0.0399403788, V: 0.0393463013, W: 0.0384232854, X: 0.0374622025, Y: 0.0368918429,
+			},
+			true: {
+				A: 0.0532814862, B: 0.0594280160, C: 0.0462076361, D: 0.0330529557, E: 0.0211688130,
+				F: 0.0594280160, G: 0.0612529898, H: 0.0462850376, I: 0.0319891593, J: 0.0213123519,
+				K: 0.0462076361, L: 0.0462850376, M: 0.0340410963, N: 0.0229646704, O: 0.0152763556,
+				P: 0.0330529557, Q: 0.0319891593, R: 0.0229646704, S: 0.0153031853, T: 0.0103681461,
+				U: 0.0211688130, V: 0.0213123519, W: 0.0152763556, X: 0.0103681461, Y: 0.0068893147,
+			},
+		},
+	},
+	{
+		g: grid(5),
+		h: func() map[int64]float64 {
+			m := make(map[int64]float64, 25)
+			for i := int64(A); i <= Y; i++ {
+				m[i] = 1
+			}
+			return m
+		}(),
+		t: 0.01, // FIXME(kortschak): Low t used due to instability in mat.Exp.
+
+		wantTol: 1e-2, // FIXME(kortschak): High tolerance used due to instability in mat.Exp.
+		want: map[bool]map[int64]float64{
+			false: {
+				A: 1, B: 1, C: 1, D: 1, E: 1,
+				F: 1, G: 1, H: 1, I: 1, J: 1,
+				K: 1, L: 1, M: 1, N: 1, O: 1,
+				P: 1, Q: 1, R: 1, S: 1, T: 1,
+				U: 1, V: 1, W: 1, X: 1, Y: 1,
+			},
+			true: {
```

Comment:
If I tighten the tolerance on this case I get the following:

```
--- FAIL: TestDiffuse (0.00s)
	diffusion_test.go:171: unexpected Diffuse result for test 3 with normalize=true:
		got: [A:0.9981753 B:1.0002977 C:0.9995572 D:1.0002977 E:0.9981753 F:1.0002977 G:1.0007715 H:1.0003861 I:1.0007715 J:1.0002977 K:0.9995572 L:1.0003861 M:1.0000019 N:1.0003861 O:0.9995572 P:1.0002977 Q:1.0007715 R:1.0003861 S:1.0007715 T:1.0002977 U:0.9981753 V:1.0002977 W:0.9995572 X:1.0002977 Y:0.9981753]
		want:[A:1.0000000 B:1.0000000 C:1.0000000 D:1.0000000 E:1.0000000 F:1.0000000 G:1.0000000 H:1.0000000 I:1.0000000 J:1.0000000 K:1.0000000 L:1.0000000 M:1.0000000 N:1.0000000 O:1.0000000 P:1.0000000 Q:1.0000000 R:1.0000000 S:1.0000000 T:1.0000000 U:1.0000000 V:1.0000000 W:1.0000000 X:1.0000000 Y:1.0000000]
```

The non-normalised operation passes. I had assumed this was because of the error in `mat.Exp`, but maybe it's not. What do you propose?

Reply:
I have run the python code equivalent to this and the overall pattern is what we see for got, but error is still there.

```
{(0, 0): 0.98264450473308107,
 (0, 1): 1.002568278028513,
 (0, 2): 0.9958911385307706,
 (0, 3): 1.002568278028513,
 (0, 4): 0.98264450473308107,
 (1, 0): 1.002568278028513,
 (1, 1): 1.0075291695232433,
 (1, 2): 1.0038067383118021,
 (1, 3): 1.0075291695232433,
 (1, 4): 1.002568278028513,
 (2, 0): 0.9958911385307706,
 (2, 1): 1.0038067383118021,
 (2, 2): 1.0001850837547184,
 (2, 3): 1.0038067383118021,
 (2, 4): 0.9958911385307706,
 (3, 0): 1.002568278028513,
 (3, 1): 1.0075291695232433,
 (3, 2): 1.0038067383118021,
 (3, 3): 1.0075291695232433,
 (3, 4): 1.002568278028513,
 (4, 0): 0.98264450473308107,
 (4, 1): 1.002568278028513,
 (4, 2): 0.9958911385307706,
 (4, 3): 1.002568278028513,
 (4, 4): 0.98264450473308107}
```